node/src/main/scala/com/wavesplatform/mining/BlockChallenger.scala

```diff
@@ -0,0 +1,277 @@
+    override def pickBestAccount(accounts: Seq[(SeedKeyPair, Long)]): Either[GenericError, (SeedKeyPair, Long)] =
+      Left(GenericError("There are no suitable accounts"))
+    override def getChallengingAccounts(challengedMiner: Address): Either[ValidationError, Seq[(SeedKeyPair, Long)]] = Right(Seq.empty)
+    override def getProcessingTx(id: ByteStr): Option[Transaction]                                                   = None
+    override def allProcessingTxs: Seq[Transaction]                                                                  = Seq.empty
+  }
+}
+
+class BlockChallengerImpl(
+    blockchainUpdater: BlockchainUpdater & Blockchain,
+    allChannels: ChannelGroup,
+    wallet: Wallet,
+    settings: WavesSettings,
+    timeService: Time,
+    pos: PoSSelector,
+    minerScheduler: Scheduler,
+    appendBlock: Block => Task[Either[ValidationError, Option[BigInt]]]
+) extends BlockChallenger
+    with ScorexLogging {
+
+  private val processingTxs: ConcurrentHashMap[ByteStr, Transaction] = new ConcurrentHashMap()
+
+  def challengeBlock(block: Block, ch: Channel, prevStateHash: ByteStr, blockReward: Option[Long]): Task[Unit] = {
+    log.debug(s"Challenging block $block")
+
+    (for {
+      _ <- EitherT.liftF(Task(processingTxs.putAll(block.transactionData.map(tx => tx.id() -> tx).toMap.asJava)))
+      challengingBlock <- EitherT(
+        createChallengingBlock(block, block.header.stateHash, block.signature, block.transactionData, prevStateHash, blockReward)
+      )
+      _ <- EitherT(appendBlock(challengingBlock).asyncBoundary)
+      _ <- EitherT.liftF[Task, ValidationError, Unit](Task(processingTxs.clear()))
+    } yield {
+      log.debug(s"Successfully challenged $block with $challengingBlock")
+      allChannels.broadcast(BlockForged(challengingBlock), Some(ch))
+    }).fold(
+      err => {
+        processingTxs.clear()
+        log.debug(s"Could not challenge $block: $err")
+      },
+      identity
+    )
+  }
+
+  def challengeMicroblock(md: MicroblockData, ch: Channel, prevStateHash: ByteStr, blockReward: Option[Long]): Task[Unit] = {
+    val idStr = md.invOpt.map(_.totalBlockId.toString).getOrElse(s"(sig=${md.microBlock.totalResBlockSig})")
+    log.debug(s"Challenging microblock $idStr")
+
+    (for {
+      discarded <- EitherT(Task(blockchainUpdater.removeAfter(blockchainUpdater.lastBlockHeader.get.header.reference)))
+      block     <- EitherT(Task(discarded.headOption.map(_._1).toRight(GenericError("Liquid block wasn't discarded"))))
+      _ <- EitherT.liftF(Task(processingTxs.putAll((block.transactionData ++ md.microBlock.transactionData).map(tx => tx.id() -> tx).toMap.asJava)))
+      challengingBlock <- EitherT(
+        createChallengingBlock(
+          block,
+          md.microBlock.stateHash,
+          md.microBlock.totalResBlockSig,
+          block.transactionData ++ md.microBlock.transactionData,
```

Comment:
Probably, we should save `block.transactionData ++ md.microBlock.transactionData` in a variable `txs` and use it twice

Reply:
Fixed